File: 069.rb
# 069
# Sqrt(x)
# 2015-12-22
#####################################################
# @param {Integer} x
# @return {Integer}
def my_sqrt(x)
  r = 1
  r = r << 1 while r*r <= x
  step = r >> 1
  while step > 0
    r -= step if (r-step)*(r-step) > x
    step = step >> 1
  end
  r-1
end

#####################################################

x = 2147483647
r = my_sqrt(x)
puts r

puts Math.sqrt(x).to_i

File: leetcode/leetcode/069.c
//
//  069.c
//  leetcode
//
//  Created by Hang Zhang on 12/23/15.
//  Copyright © 2015 Hilings Studio. All rights reserved.
//

#include <stdio.h>
#include <math.h>

int mySqrt(int x) {
    long r = 1;
    while (r*r <= x)
        r = r << 1;
    int step = (int)r >> 1;
    while (step) {
        if ((r-step)*(r-step) > x)
            r -= step;
        step = step >> 1;
    }
    return (int)r-1;
}

int main(int arg, char *argv[]) {
    // insert code here...
    printf("LeetCode 069 Sqrt(x), C ...\n\n");

    int x = 2147483647;
    int r = mySqrt(x);
    printf("%d\n", r);

    return 0;
}